File: android/src/main/java/expo/modules/mediapipepose/PoseVideoAnalysis.kt
package expo.modules.mediapipepose

import android.content.Context
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import android.os.Build
import android.os.SystemClock
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import expo.modules.kotlin.Promise
import java.util.UUID
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.atomic.AtomicBoolean

private class PoseVideoSession(
    val context: Context,
    val retriever: MediaMetadataRetriever,
    val detector: PoseLandmarker,
    val durationMs: Long,
    val options: PoseImageOptions,
) {
    val id = UUID.randomUUID().toString()
    var lastTimestamp = -1L

    fun close() {
        try {
            detector.close()
        } finally {
            retriever.release()
        }
    }
}

internal class PoseVideoAnalysis(private val masks: PoseMaskStore) {
    private val worker = Executors.newSingleThreadExecutor()
    private val destroyed = AtomicBoolean(false)
    private var session: PoseVideoSession? = null

    private fun submit(promise: Promise, operation: () -> Any?) {
        if (destroyed.get()) {
            promise.reject("videoAnalysisDisposed", "Video analysis owner is closed", null)
            return
        }
        try {
            worker.execute {
                if (destroyed.get()) {
                    promise.reject("videoAnalysisDisposed", "Video analysis owner is closed", null)
                    return@execute
                }
                try {
                    promise.resolve(operation())
                } catch (error: Exception) {
                    promise.reject(
                        "videoAnalysis",
                        "Video analysis failed: ${error.message}",
                        error,
                    )
                }
            }
        } catch (error: RejectedExecutionException) {
            promise.reject("videoAnalysisDisposed", "Video analysis owner is closed", error)
        }
    }

    fun open(
        context: Context,
        location: String,
        options: PoseImageOptions,
        trackingConfidence: Double,
        promise: Promise,
    ) =
        submit(promise) {
            check(session == null) { "Close the active video analysis before opening another" }
            require(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
                "Bounded video decoding requires Android API 27 or newer"
            }
            options.validate()
            require(trackingConfidence.isFinite() && trackingConfidence in 0.0..1.0)
            val file = PoseModel.localFile(location)
            val retriever = MediaMetadataRetriever()
            try {
                retriever.setDataSource(file.path)
                require(
                    retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_HAS_VIDEO) ==
                        "yes"
                ) {
                    "File has no video track"
                }
                val duration =
                    requireNotNull(
                            retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
                        )
                        .toLong()
                require(duration > 0)
                val configuration =
                    PoseLandmarker.PoseLandmarkerOptions.builder()
                        .setBaseOptions(PoseModel.options(options.modelVariant, options.modelPath))
                        .setRunningMode(RunningMode.VIDEO)
                        .setNumPoses(options.maxPoses)
                        .setOutputSegmentationMasks(options.segmentationEnabled)
                        .setMinPoseDetectionConfidence(options.minPoseDetectionConfidence.toFloat())
                        .setMinPosePresenceConfidence(options.minPosePresenceConfidence.toFloat())
                        .setMinTrackingConfidence(trackingConfidence.toFloat())
                        .build()
                val detector = PoseLandmarker.createFromOptions(context, configuration)
                val opened =
                    PoseVideoSession(
                        context.applicationContext,
                        retriever,
                        detector,
                        duration,
                        options,
                    )
                session = opened
                mapOf("id" to opened.id, "durationMs" to duration)
            } catch (error: Exception) {
                retriever.release()
                throw error
            }
        }

    fun read(id: String, timestampMs: Long, promise: Promise) =
        submit(promise) {
            val current = requireNotNull(session) { "Video analysis is not open" }
            require(current.id == id) { "Unknown video analysis" }
            require(
                timestampMs > current.lastTimestamp && timestampMs in 0 until current.durationMs
            )
            current.lastTimestamp = timestampMs
            try {
                check(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1)
                val limit = current.options.maxImageDimension
                val decoded =
                    requireNotNull(
                        current.retriever.getScaledFrameAtTime(
                            timestampMs * 1000,
                            MediaMetadataRetriever.OPTION_CLOSEST,
                            limit,
                            limit,
                        )
                    ) {
                        "Video frame could not be decoded"
                    }
                val pixels =
                    try {
                        requireNotNull(decoded.copy(Bitmap.Config.ARGB_8888, false))
                    } finally {
                        decoded.recycle()
                    }
                try {
                    val image = BitmapImageBuilder(pixels).build()
                    var outputMasks = emptyList<com.google.mediapipe.framework.image.MPImage>()
                    try {
                        val started = SystemClock.elapsedRealtimeNanos()
                        val result = current.detector.detectForVideo(image, timestampMs)
                        val inferenceDurationMs =
                            (SystemClock.elapsedRealtimeNanos() - started) / 1_000_000.0
                        outputMasks = result.segmentationMasks().orElse(emptyList())
                        val landmarkPayload = PoseLandmarkPayload.make(result)
                        val segmentation =
                            if (current.options.segmentationEnabled)
                                mapOf(
                                    "segmentation" to
                                        masks.save(
                                            current.context,
                                            result,
                                            pixels.width,
                                            pixels.height,
                                            current.options.maskMaxDimension,
                                        )
                                )
                            else emptyMap()
                        val detection =
                            landmarkPayload +
                                segmentation +
                                mapOf(
                                    "imageSize" to
                                        mapOf("width" to pixels.width, "height" to pixels.height),
                                    "inferenceDurationMs" to inferenceDurationMs,
                                    "model" to
                                        mapOf(
                                            "variant" to current.options.modelVariant,
                                            "delegate" to "CPU",
                                            "source" to
                                                if (current.options.modelPath == null) "bundled"
                                                else "local",
                                        ),
                                )
                        mapOf(
                            "timestampMs" to timestampMs,
                            "decodedTimestampMs" to null,
                            "detection" to detection,
                        )
                    } finally {
                        outputMasks.forEach { it.close() }
                        image.close()
                    }
                } finally {
                    if (!pixels.isRecycled) pixels.recycle()
                }
            } catch (error: Exception) {
                closeSession(id)
                throw error
            }
        }

    private fun closeSession(id: String) {
        val current = session ?: return
        if (current.id != id) return
        session = null
        current.close()
    }

    fun close(id: String, promise: Promise) =
        submit(promise) {
            closeSession(id)
            null
        }

    fun destroy() {
        if (!destroyed.compareAndSet(false, true)) return
        worker.execute {
            val current = session
            session = null
            current?.close()
        }
        worker.shutdown()
    }
}

File: ios/PoseVideoAnalysis.swift
import AVFoundation
import MediaPipeTasksVision
import UIKit

private final class PoseVideoSession {
  enum Phase { case ready, decoding }
  let id = UUID().uuidString
  let generator: AVAssetImageGenerator
  let durationMs: Double
  let options: PoseImageOptions
  var detector: PoseLandmarker?
  var lastTimestamp = -1
  var phase = Phase.ready

  init(asset: AVAsset, durationMs: Double, options: PoseImageOptions, trackingConfidence: Double)
    throws
  {
    self.durationMs = durationMs
    self.options = options
    generator = AVAssetImageGenerator(asset: asset)
    generator.appliesPreferredTrackTransform = true
    generator.maximumSize = CGSize(
      width: options.maxImageDimension, height: options.maxImageDimension)
    generator.requestedTimeToleranceBefore = .zero
    generator.requestedTimeToleranceAfter = .zero
    let configuration = PoseLandmarkerOptions()
    configuration.baseOptions.modelAssetPath = try PoseModel.path(
      variant: options.modelVariant, localPath: options.modelPath)
    configuration.baseOptions.delegate = .CPU
    configuration.runningMode = .video
    configuration.numPoses = options.maxPoses
    configuration.minPoseDetectionConfidence = Float(options.minPoseDetectionConfidence)
    configuration.minPosePresenceConfidence = Float(options.minPosePresenceConfidence)
    configuration.minTrackingConfidence = Float(trackingConfidence)
    detector = try PoseLandmarker(options: configuration)
  }

  func close() {
    generator.cancelAllCGImageGeneration()
    detector = nil
  }
}

internal final class PoseVideoAnalysis {
  let worker = DispatchQueue(label: "expo.pose.media")
  private var session: PoseVideoSession?
  private var destroyed = false

  private func onWorker<Value>(_ operation: @escaping () throws -> Value) async throws -> Value {
    try await withCheckedThrowingContinuation { continuation in
      worker.async {
        do { continuation.resume(returning: try operation()) } catch {
          continuation.resume(throwing: error)
        }
      }
    }
  }

  func open(_ location: String, options: PoseImageOptions, trackingConfidence: Double) async throws
    -> [String: Any]
  {
    try options.validate()
    guard trackingConfidence.isFinite, (0...1).contains(trackingConfidence) else {
      throw PoseMediaError.invalidOptions
    }
    let asset = AVURLAsset(url: try PoseModel.localURL(location))
    let duration = try await asset.load(.duration)
    let tracks = try await asset.loadTracks(withMediaType: .video)
    let durationMs = duration.seconds * 1000
    guard !tracks.isEmpty, durationMs.isFinite, durationMs > 0 else {
      throw PoseMediaError.invalidFile
    }
    return try await onWorker {
      guard !self.destroyed, self.session == nil else { throw PoseMediaError.videoUnavailable }
      let opened = try PoseVideoSession(
        asset: asset, durationMs: durationMs, options: options,
        trackingConfidence: trackingConfidence)
      self.session = opened
      return ["id": opened.id, "durationMs": durationMs]
    }
  }

  func read(_ id: String, timestampMs: Int) async throws -> [String: Any] {
    let generator: AVAssetImageGenerator = try await onWorker {
      guard let current = self.session, current.id == id else {
        throw PoseMediaError.videoUnavailable
      }
      guard current.phase == .ready, timestampMs > current.lastTimestamp else {
        throw PoseMediaError.invalidOptions
      }
      guard timestampMs >= 0, Double(timestampMs) < current.durationMs else {
        throw PoseMediaError.invalidOptions
      }
      current.lastTimestamp = timestampMs
      current.phase = .decoding
      return current.generator
    }
    do {
      let decoded = try await generator.image(
        at: CMTime(value: Int64(timestampMs), timescale: 1000))
      return try await onWorker {
        try autoreleasepool {
          guard let current = self.session, current.id == id, let detector = current.detector else {
            throw PoseMediaError.videoUnavailable
          }
          let image = try MPImage(uiImage: UIImage(cgImage: decoded.image))
          let started = ProcessInfo.processInfo.systemUptime
          let result = try detector.detect(videoFrame: image, timestampInMilliseconds: timestampMs)
          var detection = try PoseLandmarkPayload.make(result)
          detection["imageSize"] = ["width": decoded.image.width, "height": decoded.image.height]
          detection["inferenceDurationMs"] = (ProcessInfo.processInfo.systemUptime - started) * 1000
          detection["model"] = [
            "variant": current.options.modelVariant, "delegate": "CPU",
            "source": current.options.modelPath == nil ? "bundled" : "local",
          ]
          current.phase = .ready
          return [
            "timestampMs": timestampMs, "decodedTimestampMs": decoded.actualTime.seconds * 1000,
            "detection": detection,
          ]
        }
      }
    } catch {
      try await onWorker { self.close(id) }
      throw error
    }
  }

  func close(_ id: String) {
    guard let current = session, current.id == id else { return }
    current.close()
    session = nil
  }

  func destroy() {
    destroyed = true
    session?.close()
    session = nil
  }
}
